crates/circuit/src/parameter/parameter_expression.rs

```diff
@@ -0,0 +1,2162 @@
+            _ => PyRuntimeError::new_err(value.to_string()),
+        }
+    }
+}
+
+/// A parameter expression.
+///
+/// This is backed by Qiskit's symbolic expression engine and a cache
+/// for the parameters inside the expression.
+#[derive(Clone, Debug)]
+pub struct ParameterExpression {
+    // The symbolic expression.
+    pub expr: SymbolExpr, // TODO make these attrs private?
+    // A map keeping track of all symbols, with their name. This map *must* be kept
+    // up to date upon any operation performed on the expression.
+    pub name_map: HashMap<String, Symbol>,
+}
+
+impl Hash for ParameterExpression {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.expr.to_string().hash(state);
+    }
+}
+
+impl PartialEq for ParameterExpression {
+    fn eq(&self, other: &Self) -> bool {
+        self.expr.eq(&other.expr)
+    }
+}
+
+impl Eq for ParameterExpression {}
+
+impl Default for ParameterExpression {
+    /// The default constructor returns zero.
+    fn default() -> Self {
+        Self {
+            expr: SymbolExpr::Value(Value::Int(0)),
+            name_map: HashMap::new(), // no parameters, hence empty name map
+        }
+    }
+}
+
+impl fmt::Display for ParameterExpression {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", {
+            if let SymbolExpr::Symbol(s) = &self.expr {
+                s.name()
+            } else {
+                match self.expr.eval(true) {
+                    Some(e) => e.to_string(),
+                    None => self.expr.optimize().to_string(),
+                }
+            }
+        })
+    }
+}
+
+/// Lookup for which operations are binary (i.e. require two operands).
+static BINARY_OPS: [OpCode; 8] = [
+    // a HashSet would be better but requires unstable features
+    OpCode::ADD,
+    OpCode::SUB,
+    OpCode::MUL,
+    OpCode::DIV,
+    OpCode::POW,
+    OpCode::RSUB,
+    OpCode::RDIV,
+    OpCode::RPOW,
+];
+
+impl ParameterExpression {
+    /// Initialize with an existing [SymbolExpr] and its valid name map.
+    ///
+    /// Caution: The caller **guarantees** that ``name_map`` is consistent with ``expr``.
+    /// If uncertain, call [Self::from_symbol_expr], which automatically builds the correct name map.
+    pub fn new(expr: &SymbolExpr, name_map: &HashMap<String, Symbol>) -> Self {
+        Self {
+            expr: expr.clone(),
+            name_map: name_map.clone(),
+        }
+    }
+
+    /// Construct from a [Symbol].
+    pub fn from_symbol(symbol: &Symbol) -> Self {
+        Self {
+            expr: SymbolExpr::Symbol(symbol.clone()),
+            name_map: [(symbol.name(), symbol.clone())].into(),
+        }
+    }
+
+    /// Construct from a [SymbolExpr].
+    ///
+    /// This populates the name map with the symbols in the expression.
+    pub fn from_symbol_expr(expr: SymbolExpr) -> Self {
+        Self {
+            expr: expr.clone(),
+            name_map: expr.name_map().clone(),
+        }
+    }
```

Comment:
Do like this 🙂 

```suggestion
    pub fn from_symbol_expr(expr: SymbolExpr) -> Self {
        let name_map = expr.name_map();
        Self {
            expr,
            name_map,
        }
    }
```

Otherwise you've got extra clones you certainly don't want. In your current version, I think the `.clone()` on `name_map()` is actually fully unnecessary anyhow!

Reply:
ah that's much better, thanks !